### PR TITLE
[iOS] RecycleElement ListView should not crash when bound to ObservableCollection (ArgumentNullException & ArgumentException)

### LINKED ISSUE #1900: Xamarin ios ListView ObservableCollection<myClass>. Collection.Add() throwing 'Index # is greater than the number of rows #' exception
### Description
Xamarin ios ListView ObservableCollection<myClass>. Collection.Add() throwing 'Index "11" is greater than the number of rows "10"' exception. 

Not using groups - found bugzilla report [here](https://bugzilla.xamarin.com/show_bug.cgi?id=59896) supposedly fixed in my version.

 

```
private ListView ItemList { get; set; }
        private ObservableCollection<MyObject> ItemSource { get; set; }
        ItemList = new ListView(ListViewCachingStrategy.RecycleElement)
            {
                ItemTemplate = new DataTemplate(() => { ... }),
                ...
            };
            ...
            ItemSource = new ObservableCollection<MyObject>();
            ItemList.ItemsSource = ItemSource;


        public void AddItemsToList(IEnumerable<Object> items)
        {
            foreach (var item in items)
            {
                ItemSource.Add(new ObjectContainer() { Object = item });
            }
        }
```

Stack Trace
```
  at Xamarin.Forms.Platform.iOS.ListViewRenderer.UpdateItems (System.Collections.Specialized.NotifyCollectionChangedEventArgs e, System.Int32 section, System.Boolean resetWhenGrouped) [0x0006e] in D:\agent\_work\1\s\Xamarin.Forms.Platform.iOS\Renderers\ListViewRenderer.cs:528 
  at Xamarin.Forms.Platform.iOS.ListViewRenderer.OnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00000] in D:\agent\_work\1\s\Xamarin.Forms.Platform.iOS\Renderers\ListViewRenderer.cs:307 
  at Xamarin.Forms.Internals.TemplatedItemsList`2[TView,TItem].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x0000a] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.Internals.TemplatedItemsList`2[TView,TItem].OnProxyCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e, System.Boolean fixWindows) [0x0001a] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.Internals.TemplatedItemsList`2[TView,TItem].OnProxyCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00000] in <1aabc0fecdaa41b6a5c04760c77bb648>:0 
  at Xamarin.Forms.ListProxy.OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x0000a] in D:\agent\_work\1\s\Xamarin.Forms.Core\ListProxy.cs:233 
  at Xamarin.Forms.ListProxy+<>c__DisplayClass33_0.<OnCollectionChanged>b__0 () [0x00018] in D:\agent\_work\1\s\Xamarin.Forms.Core\ListProxy.cs:206 
  at Xamarin.Forms.ListProxy.OnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x000a0] in D:\agent\_work\1\s\Xamarin.Forms.Core\ListProxy.cs:225 
  at Xamarin.Forms.ListProxy+WeakNotifyProxy.OnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00031] in D:\agent\_work\1\s\Xamarin.Forms.Core\ListProxy.cs:394 
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x0000f] in <7e690ac38a6a43829ef37bc1fa194d96>:0 
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedAction action, System.Object item, System.Int32 index) [0x00009] in <7e690ac38a6a43829ef37bc1fa194d96>:0 
  at System.Collections.ObjectModel.ObservableCollection`1[T].InsertItem (System.Int32 index, T item) [0x00024] in <7e690ac38a6a43829ef37bc1fa194d96>:0 
  at System.Collections.ObjectModel.Collection`1[T].Add (T item) [0x00020] in <d680a8bf3a2a4c8dbbfde665ba81f6f2>:0 
  at MyApp.Controls.Products.ItemListView.AddItemsToList (System.Collections.Generic.IEnumerable`1[T] items) [0x00014] in C:\GitRepos\MyApp\Controls\Products\ItemListView.cs:85 
```

### Steps to Reproduce

1.  Create item list and set it's ItemSource to ObservableCollection
2.  Add items to Observable collection in IOS

### Expected Behavior
observableCollection allows me to add new items into it. Works in Android

### Actual Behavior
Throws exception on second item in loop. 
Runs through the first 10 items fine, but when I add the next 10, it loops through the first item in that collection, and throws the exception on the second.

### Basic Information

- Version with issue: - Include="Xamarin.Forms" Version="2.5.0.122203"
- Last known good version: - New issue
- IDE: - Visual Studio
- Platform Target Frameworks:
  - iOS: 9.0 - attempting to run on 11.2
  - Android: Works fine
- Nuget Packages: 
    - CarouselView.FormsPlugin Version="5.2.0"
    - Newtonsoft.Json" Version="10.0.3"
    - Rg.Plugins.Popup" Version="1.1.4.145-pre"
    - Xamarin.FFImageLoading.Forms" Version="2.3.4"
- Affected Devices:
   - IOS emulator and iPhone 6 (real device)

### Screenshots

image

### LINKED ISSUE #3026: [iOS] ListView throws ArgumentNullException when HasUnevenRows = true and no ItemTemplate specified
### Description

Using a ListView with no ItemTemplate specified (i.e., defaults to a `TextCell` with `HasUnevenRows` causes a crash when attempting to cache the prototypical cell because the key it tries to store it as is null.

Related to #1143

### Steps to Reproduce

```csharp
public class ListPage : ContentPage
		{
			public ListPage()
			{
				var listView = new ListView(ListViewCachingStrategy.RecycleElement) { AutomationId = "ListView", HasUnevenRows = true, ItemsSource = Enumerable.Range(0, 10).Select(i => $"Item {i}") };
				Content = listView;
			}
		}
```

### Expected Behavior

Show 10 items in a list

### Actual Behavior

Crashes: `ArgumentNullException: parameter name key`

### Basic Information

- Version with issue: 3.1.0
- Last known good version: ? probably 2.4.0
- Platform Target Frameworks: 
  - iOS: 11.3

### PR DESCRIPTION
**Pull request by samhouts**
### Description of Change ###

For #1900, an exception was being thrown inappropriately when the ListView used RecycleElement/AndDataTemplate caching strategies. This exception is now skipped in that case.

For #3026, an exception was thrown when a ListView with HasUnevenRows and no ItemTemplate was specified. Now informing that code that the the ItemTemplate defaults to TextCell for purposes of caching the ItemTemplate.

### Issues Resolved ###

- fixes #1900 
- fixes #3026

### API Changes ###

None

### Platforms Affected ###

- iOS

### Behavioral/Visual Changes ###

None

### PR Checklist ###

- [x] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 999999, "Xamarin ios ListView ObservableCollection<myClass>. Collection.Add() throwing 'Index # is greater than the number of rows #' exception", PlatformAffected.iOS)]
```
Comment (rmarinho) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs]: replace the 99999 ?

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+			{
+				Items.Add(item);
+			}
+		}
+
+		protected override void Init()
+		{
+			var listView = new ListView(ListViewCachingStrategy.RecycleElement) { AutomationId = "ListView", ItemsSource = Items };
+			listView.ItemAppearing += ItemList_ItemAppearing;
+			Content = new StackLayout { Children = { new Label { Text = "If this test crashes when i loads or when you scroll the list, then this test has failed. Obviously." }, listView } };
```
Comment (hartez) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs]: "when i loads" -> "when it loads"

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+using Xamarin.Forms.Internals;
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
```
Comment (hartez) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs]: This needs `UITestCategories.ListView`.

Comment (samhouts) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue1900.cs]: Done!

**Pull request merged**